### PR TITLE
Add some more functions to algorithm.hpp

### PR DESCRIPTION
**Pull request by Sebanisu**
I wanted to be able to test `constexpr` versions of the functions. So i'm including `<array>` for testing and added some range based overloads.
Added predicates:
`equal_to`
`not_equal_to`
`is_equal` - is_equal is a variadic version of `std::equal`
Added range overloads in `ranges` namespace.
`zip_find`
`zip_found`
`is_equal`

`is_mismatched` was removed as `!std::equal` covers it.

`auto` has to be on the right of the cv qualifiers. Or the clang-format new line won't trigger.

### CODE REVIEW

**Comment by Sebanisu** [comment]
I thought you could pass an r-value to a const ref. Though it's saying you can't.  I mean I thought it would implicitly convert.

**Comment by Sebanisu** [comment]
oh i forgot const

**Comment by Sebanisu** [comment]
I was thinking that the end iterator should be const. Though at work cannot fix rn.

File: jsrc/algorithm.hpp [hunk 1]
```diff
@@ -1,39 +1,119 @@
...
 #include <algorithm>
+#include <array>
 #include <tuple>
 
 namespace algo {
 
+/**
+ * @brief Variadic version of `std::equal_to`
+ */
+static auto constexpr is_equal_to = [](auto const &first, auto const &...values) -> bool {
```
Comment (codereport) [inline jsrc/algorithm.hpp:10]:
```suggestion
static auto constexpr equal_to = [](auto const &first, auto const &...values) -> bool {
```

File: jsrc/algorithm.hpp [hunk 2]
```diff
@@ -1,39 +1,119 @@
...
+    static_assert(sizeof...(values) != 0);
+    return ((first == values) && ...);
+};
+
+/**
+ * @brief Variadic version of `std::not_equal_to`
+ */
+static auto constexpr is_not_equal_to = [](auto const &first, auto const &...values) -> bool {
```
Comment (codereport) [inline jsrc/algorithm.hpp:18]:
```suggestion
static auto constexpr not_equal_to = [](auto const &first, auto const &...values) -> bool {
```

File: jsrc/algorithm.hpp [hunk 3]
```diff
@@ -1,39 +1,119 @@
...
     }
     return std::tuple{f, fs...};
 }
 
+/**
+ * @brief Variadic version of `std::mismatch`, for ranges.
+ * @note This can be the same name as zip_find if we use concepts.
+ */
+template <typename P, typename R, typename... Rs>
+[[nodiscard]] static auto constexpr zip_find_r(P const &pred, R const &r, Rs const &...rs) {
```
Comment (codereport) [inline jsrc/algorithm.hpp:40]: Let's just put this in a nested namespace and drop the `_r`. `algo::ranges::` probably to be consistent with C++ Ranges.

File: jsrc/algorithm.hpp [hunk 4]
```diff
@@ -1,39 +1,119 @@
...
+[[nodiscard]] static auto
+is_mismatched(I f, I l, I2 f2) -> bool {
     return std::mismatch(f, l, f2).first != l;
 }
 
+/**
+ * @brief Variadic, Predicate version of `std::mismatch`
+ */
+template <typename I, typename... Is>
+[[nodiscard]] static auto constexpr is_mismatched_v(I f, I l, Is... fs) -> bool {
```
Comment (codereport) [inline jsrc/algorithm.hpp:76]: looks like file needs to be clang-formatted.

Comment (codereport) [inline jsrc/algorithm.hpp:76]: Don't think we need `_v`, function overloading is fine.
```suggestion
[[nodiscard]] static auto constexpr is_mismatched(I f, I l, Is... fs) -> bool {
```

**Review by codereport — CHANGES_REQUESTED** [review]

File: jsrc/algorithm.hpp (hunk 4 above)
Comment (Sebanisu) [reply jsrc/algorithm.hpp:76]: I just wasn't sure if I should replace the existing one.

Comment (Sebanisu) [reply jsrc/algorithm.hpp]: I ran formatting on my pc I wonder if it's not parsing the clang.format file.

**Comment by Sebanisu** [comment]
I'll finish the requested changes on about 6 hours 👍

File: jsrc/algorithm.hpp (hunk 4 above)
Comment (Sebanisu) [reply jsrc/algorithm.hpp]: I'm a little confused .clang_format seems to be working. Though this line doesn't change.

Comment (juntuu) [reply jsrc/algorithm.hpp]: Seems like clang format doesn't break after `static` or `constexpr`

so either `auto` should be last one before the name, or clang format needs configuring, (or the code needs manual formatting adjustment)

Comment (Sebanisu) [reply jsrc/algorithm.hpp]: okay so I put auto on the right and it does the newline. So i'll mark this resolved. There might be a way to new line after cv qualifiers.

File: jsrc/algorithm.hpp [hunk 5]
```diff
@@ -21,19 +37,108 @@ zip_find(P pred, I f, I l, Is... fs) {
...
+// cpp20.
+// /**
+//  * @brief Predicate version of `std::mismatch`
+//  * @note std::mismatch isn't constexpr till cpp20
+//  */
+// template <typename I, typename I2>
+// [[nodiscard]] static auto
+// is_mismatched(I f, I const l, I2 f2) -> bool {
+//     return std::mismatch(f, l, f2).first != l;
+// }
```
Comment (codereport) [inline jsrc/algorithm.hpp:56]:
```suggestion
```

File: jsrc/algorithm.hpp [hunk 6]
```diff
@@ -21,19 +37,108 @@ zip_find(P pred, I f, I l, Is... fs) {
...
+// cpp20.
+// /**
+//  * @brief Predicate version of `std::equal`
+//  * @note std::equal isn't constexpr till cpp20
+//  */
+// template <typename I, typename I2>
+// [[nodiscard]] static auto
+// is_equal(I f, I const l, I2 f2) -> bool {
+//     return std::equal(f, l, f2);
+// }
```
Comment (codereport) [inline jsrc/algorithm.hpp:77]:
```suggestion
```

File: jsrc/algorithm.hpp [hunk 7]
```diff
@@ -21,19 +37,108 @@ zip_find(P pred, I f, I l, Is... fs) {
...
+
+/**
+ * @brief Variadic, Predicate version of `std::mismatch`
+ */
+template <typename I, typename... Is>
+[[nodiscard]] static constexpr auto
+is_mismatched(I f, I const l, Is... fs) -> bool {
+    return zip_found(not_equal_to, f, l, fs...);
+}
```
Comment (codereport) [inline jsrc/algorithm.hpp:65]: Realized `is_mismatched` is the negation of `std::equal`, so lets just rework this one and remove `is_mismatched`.
```suggestion
is_equal(I f, I const l, Is... fs) -> bool {
    return zip_found(equal_to, f, l, fs...);
}
```

File: jsrc/algorithm.hpp [hunk 8]
```diff
@@ -21,19 +37,108 @@ zip_find(P pred, I f, I l, Is... fs) {
...
+ * @brief Variadic, Predicate version of `std::equal`, for ranges;
  */
-template <typename I, typename I2>
-[[nodiscard]] auto
-is_mismatched(I f, I l, I2 f2) {
-    return std::mismatch(f, l, f2).first != l;
+template <typename... Rs>
+[[nodiscard]] static constexpr auto
+is_equal(Rs const &...rs) -> bool {
+    return !is_mismatched(rs...);
```
Comment (codereport) [inline jsrc/algorithm.hpp:126]: :heart:

**Review by codereport — CHANGES_REQUESTED** [review]
this looks really good.

**Review by codereport — APPROVED** [review]
:fire:

**Pull request merged**